feat: Add knb plugin build tool

dsimansk opened the pull request:
## Description

This PR adds [knb](https://github.com/dsimansk/knb) tool from temp repository.

It should allow the binary to be installable via go get, e.g.:
```
go get knative.dev/client/knb
```


## Changes

<!-- Please add list of more detailed changes. These changes should be reflected also in the commit messages -->

* Add `knb` module
* Changelog `TBD`.

## Reference

<!-- Please add the corresponding issue number which this pull request is about to fix -->
Follow-up https://github.com/knative/client/issues/928#issuecomment-661900519

/lint

<!--
Please add an entry to CHANGELOG.adoc file, too, as part of your Pull Request.

In the following cases, add a short description of PR to the unreleased section in CHANGELOG.adoc:

- 🎁 New feature
- 🐛 Bug fix
- ✨ Feature Update
- 🐣 Refactoring
- 🗑️ Remove feature or internal logic

See other entries in CHANGELOG.adoc as an example for how to add the entry, including a reference to the corresponding issue/PR

PLEASE DON'T ADD THAT LINE HERE IN THE PULL-REQUEST DESCRIPTION BUT DIRECTLY IN CHANGELOG.ADOC AND ADD CHANGELOG.ADOC AS PART OF YOUR PULL-REQUEST.
-->

<!--
To automatically lint go code in this pull request uncomment the line below. You get feedback as comments on your pull request then -->

<!--
/lint
-->

dsimansk commented:
/lint

rhuss reviewed (commented):
I just detected https://github.com/golang-standards/project-layout and they recommedn to put tools below, well, `tools/` ;-) Maybe you could move knb into a tools, directory, too ? Also wonder whether we should integrate it into the regular build in hack/build.sh or use a dedicated build process for knb (guess not everyone wants to build it everytime, especially as I expect it to grow a bit).

For CI, we could delegate it from the top-level scripts and e.g. check whether a knb file was touched (I had such code in the client-contrib repo to detect if a PR touches a certain sub dir).

dsimansk commented:
> I just detected https://github.com/golang-standards/project-layout and they recommedn to put tools below, well, `tools/` ;-) Maybe you could move knb into a tools, directory, too ? Also wonder whether we should integrate it into the regular build in hack/build.sh or use a dedicated build process for knb (guess not everyone wants to build it everytime, especially as I expect it to grow a bit).

I see, it sound good to me. The desired structure would be `./tools/knb/knb.go`?


> 
> For CI, we could delegate it from the top-level scripts and e.g. check whether a knb file was touched (I had such code in the client-contrib repo to detect if a PR touches a certain sub dir).


I've added knb to `sources_dir` variable only to benefit from code common formatting, the variable itself doesn't seem to be used for build etc.[[1](https://github.com/knative/client/blob/master/hack/build.sh#L108-L111)][[2](https://github.com/knative/client/blob/master/hack/build.sh#L130-L135)] In other words formatting should be safe, but I don't adding a bit of "isChanged" magic.

rhuss commented:
> I see, it sound good to me. The desired structure would be `./tools/knb/knb.go`?

yep

> Would you like to follow dir naming `/internal` vs. `/pkg`?

I think we can do this in a later step, as it's not really clear what we want to expose on that level. Depends also on what we want to put into the shared repository,

maximilien commented on tools/knb/README.md:
```diff
@@ -0,0 +1,112 @@
+
+##### Create custom `kn` distribution
+
+The `knb` can be used to generate enhanced `kn` source files with inlined plugins.
```
I would say: `enhanced customized ...`

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/a06c275eb3d0ae388efb1441679cb36444c5f741.

maximilien commented on tools/knb/README.md:
```diff
@@ -0,0 +1,112 @@
+
+The `knb` can be used to generate enhanced `kn` source files with inlined plugins.
+
+Create configuration file `.kn.yaml` in a root directory of `knative/client` that should specify at least `name, module, version` coordinates of the plugin.
```
Maybe show a sample YAML inline here to make it easy for user to see and cut and paste

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/a06c275eb3d0ae388efb1441679cb36444c5f741.

maximilien commented on tools/knb/README.md:
```diff
@@ -0,0 +1,112 @@
+The `knb` can be used to generate enhanced `kn` source files with inlined plugins.
+
+Create configuration file `.kn.yaml` in a root directory of `knative/client` that should specify at least `name, module, version` coordinates of the plugin.
+The `plugin distro` will generated the required go files and add dependency to `go.mod`.
```
Not clear what the `plugin distro` is?

maximilien replied:
I know it’s a command of the `knb` tool but not clear here

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/a06c275eb3d0ae388efb1441679cb36444c5f741.

maximilien commented on tools/knb/README.md:
````diff
@@ -0,0 +1,112 @@
+* replace - go module replacement defined by `module,version`.
+
+Example of `.kn.yaml`
+```yaml
````
Move this near the first discussion above

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/a06c275eb3d0ae388efb1441679cb36444c5f741.

maximilien commented on tools/knb/README.md:
```diff
@@ -0,0 +1,112 @@
+    pluginImportPath: knative.dev/kn-plugin-source-kafka/plugin
+    version: v0.19.0
+    replace:
+      - module: foo.bar
```
Not clear what this is since the example is using `foo.bar` but the plugin is a real one...

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/a06c275eb3d0ae388efb1441679cb36444c5f741.

maximilien commented on tools/knb/README.md:
````diff
@@ -0,0 +1,112 @@
+The `knb` can be used to generate required go files to inline any `kn` plugin.
+
+```bash
+knb plugin init --name kn-source-kafka --cmd source,kafka --description "Some plugin"
````
Why not link to the generated docs for `knb` so that the options are in the natural expected fashion. For instance `—cmd` is an option but not clear what the possible values are...

dsimansk replied:
It was meant to represent one line example. The cmd docs are added below, but I can add the same generation of cmd descriptions to `./docs` dir of course.

maximilien commented on tools/knb/pkg/plugin/distro.go:
```diff
@@ -0,0 +1,159 @@
+			if _, err := os.Stat(config); os.IsNotExist(err) {
+				return fmt.Errorf("kn distro configuration file '%s' doesn't exist", config)
+			}
+			if _, err := os.Stat("cmd/kn/main.go"); os.IsNotExist(err) {
```
I would move all these file checks into a utility function that you can call here in sequence for each required file.

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien commented on tools/knb/pkg/plugin/distro.go:
```diff
@@ -0,0 +1,159 @@
+			fmt.Println("✔  config file '" + config + "' processed")
+
+			for _, p := range conf.Plugins {
+				importPath := p.PluginImportPath
```
Can we also simplify this for loop by having the module processing into it’s own function?

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien commented on tools/knb/pkg/plugin/distro.go:
```diff
@@ -0,0 +1,159 @@
+	return ioutil.WriteFile(file, content, 0644)
+}
+
+var registerTemplate = `
```
Could this be a constant? Also does it make sense to have the var into its own file to make it easy for this file to read and reduce the size?

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien commented on tools/knb/pkg/plugin/distro.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+var registerTemplate = `
+// Copyright © 2020 The OpenShift Knative Authors
```
`OpenShift` ?

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien commented on tools/knb/pkg/plugin/init.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+var pluginTemplate = `
+package plugin
```
Same as before. I would have them in their own `plugin_template.go` that exports the const var or similar

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien commented on tools/knb/pkg/plugin/plugin.go:
```diff
@@ -0,0 +1,44 @@
+
+// Plugin represents plugin configuration data struct
+type Plugin struct {
+	Name             string
```
Why not be explicit about the YAML annotation so it’s clear. Things like `PluginImportPath` make it hard to guess what the YAML field is?

dsimansk replied:
Fixed in https://github.com/knative/client/pull/1226/commits/d940570529633941f4fc8e88d98bea95f35d3b14.

maximilien reviewed (approved):
Important contribution and great start. Thanks.

Left various feedback to helpfully make code more modular and clearer.

/ok-to-test

dsimansk commented:
Hi, @maximilien! Thanks for the feedback. I've hopefully address all of the comments. In addition we have a newly enabled `gosec` that spotted a few issue I'll fix them soon.

maximilien commented:
> Hi, @maximilien! Thanks for the feedback. I've hopefully addressed all of the comments. In addition we have a newly enabled `gosec` that spotted a few issue I'll fix them soon.

Cool. Let me know when ready for final review

dsimansk commented:
@maximilien I've resolved the addressed review comments. There's one left I'd tackle it as an improvement to the docs, examples etc as suggested during WG call. I'm going to open an issue to track it.
Otherwise it's good for final review.

rhuss reviewed (commented):
Looks good, but if you don't mind let's wait for 0.21 release performed first ?

dsimansk commented:
@rhuss @maximilien please let me know if there's anything outstanding to be added to this PR or can be deferred to future iterations. Thanks!

rhuss reviewed (approved):
thanks !

/lgtm
/apply
